File: api/services/redis.py
import json, os
from typing import Any, Optional
import redis.asyncio as aioredis

REDIS_URL = os.getenv("REDIS_URL", "redis://localhost:6379/0")
CACHE_TTL = int(os.getenv("CACHE_TTL", "3600"))

_client: Optional[aioredis.Redis] = None

async def get_redis() -> aioredis.Redis:
    global _client
    if _client is None:
        _client = await aioredis.from_url(REDIS_URL, encoding="utf-8", decode_responses=True)
    return _client

async def redis_get(key: str) -> Optional[Any]:
    r = await get_redis()
    raw = await r.get(key)
    return json.loads(raw) if raw else None

async def redis_set(key: str, value: Any, ttl: int = CACHE_TTL) -> None:
    r = await get_redis()
    await r.setex(key, ttl, json.dumps(value, default=str))

async def redis_del(key: str) -> bool:
    r = await get_redis()
    return bool(await r.delete(key))

async def close():
    global _client
    if _client:
        await _client.aclose()
        _client = None

File: api/services/redis.js
import Redis from "ioredis";

let client = null;

export function getRedis() {
  if (!client) {
    client = new Redis(process.env.REDIS_URL || "redis://localhost:6379/0", {
      lazyConnect: true,
      retryStrategy: (times) => Math.min(times * 100, 3000),
    });
    client.on("error", (err) => console.error("[Redis]", err.message));
  }
  return client;
}

export async function redisGet(key) {
  try {
    const raw = await getRedis().get(key);
    return raw ? JSON.parse(raw) : null;
  } catch { return null; }
}

export async function redisSet(key, value, ttl = Number(process.env.CACHE_TTL) || 3600) {
  try {
    await getRedis().setex(key, ttl, JSON.stringify(value));
  } catch (e) { console.error("[Redis set]", e.message); }
}

export async function redisDel(key) {
  try { return await getRedis().del(key); } catch { return 0; }
}

export async function redisKeys(pattern) {
  try { return await getRedis().keys(pattern); } catch { return []; }
}

export async function closeRedis() {
  if (client) { await client.quit(); client = null; }
}
